auth/auth.go

```diff
@@ -0,0 +1,172 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// UserIsLoggedIn will read the JWT in the request header and verify that it is legitimate.
+func (a *Auth) UserIsLoggedIn() gin.HandlerFunc {
+	return func(c *gin.Context) {
+
+		token, err := a.getToken(c)
+
+		if err != nil {
+			c.Abort()
+			return
+		}
+
+		if token.Valid {
+			// save token in context for use in other middleware
+			c.Set("token", token)
+		} else {
+			origin := c.GetHeader("Origin")
+			if ve, ok := err.(*jwt.ValidationError); ok {
+				if ve.Errors&jwt.ValidationErrorMalformed != 0 {
+					c.JSON(http.StatusForbidden, gin.H{"err": "token is malformed"})
+					log.Error(fmt.Sprintf("Attempt to access resources with malformed token from origin: %s", origin))
+				} else if ve.Errors&(jwt.ValidationErrorExpired) != 0 {
+					c.JSON(http.StatusForbidden, gin.H{"err": "token is expired"})
+				} else {
+					c.JSON(http.StatusForbidden, gin.H{"err": "error reading token"})
+					log.Error(fmt.Sprintf("Error accessing resources with token from origin %s", origin))
+				}
+			} else {
+				c.JSON(http.StatusForbidden, gin.H{"err": "error reading token"})
+				log.Error(fmt.Sprintf("Error accessing resources with token from origin %s", origin))
+			}
+			c.Abort()
+		}
+	}
+}
+
+// UserIsInRole accepts a list of roles and determines whether a user's role in a JWT is in that list.
+func (a *Auth) UserIsInRole(d database.IDatabase, roles []constants.Role) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		var token interface{}
+		var ok bool
+
+		if token, ok = c.Get("token"); !ok {
+			c.JSON(http.StatusForbidden, gin.H{"err": "Error accessing token"})
+			c.Abort()
+			return
+		}
+
+		var userRole constants.Role
+		if claims, ok := token.(*jwt.Token).Claims.(jwt.MapClaims); ok {
+			userRole = constants.Role(claims["role"].(float64))
+		} else {
+			c.Abort()
+			return
+		}
+
+		for _, role := range roles {
+			if userRole == role {
+				return
+			}
+		}
+
+		c.JSON(http.StatusForbidden, gin.H{"err": "User doesn't have access"})
+		c.Abort()
+	}
+}
+
+// CreateToken generates a signed JWT string based on user info.
+func (a *Auth) CreateToken(user model.User) (string, error) {
+	token := jwt.New(jwt.SigningMethodRS256)
+	claims := make(jwt.MapClaims)
+	claims["exp"] = time.Now().Add(time.Hour * 24 * 7).Unix() // expires in one week
+	claims["iat"] = time.Now().Unix()
+	claims["user"] = user.Username
+	claims["id"] = user.ID
+	claims["role"] = user.UserRole
+	token.Claims = claims
+
+	return token.SignedString(signKey)
+}
+
+func (a *Auth) getToken(c *gin.Context) (*jwt.Token, error) {
+	tokenString := c.GetHeader("Authorization")
+
+	if tokenString == "" {
+		origin := c.GetHeader("Origin")
+		c.JSON(http.StatusForbidden, gin.H{"err": "token is required"})
+		log.Error(fmt.Sprintf("Attempt to access resources without token from origin: %s", origin))
+		return nil, errors.New("Error")
+	}
+
+	tokenString = tokenString[len("Bearer "):]
```

Comment:
This reads weird to me, not sure why?

Reply:
Yeah this is kinda weird, think I'll just replace with a strings.Replace()